extensions/metadata.md

```diff
@@ -0,0 +1,740 @@
+
+*Failures*: `FAIL METADATA KEY_NO_PERMISSION`
+
+### METADATA SUB
+
+    METADATA * SUB <key1> [<key2> ...]
+
+This subcommand lets clients subscribe to updates for the given keys.
+
+The server MUST process each key in order, as the client uses this order to determine which keys they were and were not able to subscribe to. Clients SHOULD send keys in order of preference.
+
+The server processes each key in order, and:
+
+> If the client has subscribed to too many keys, the server does not process any further keys in the subcommand. The `<key>` parameter of the `FAIL METADATA TOO_MANY_SUBS` reply MUST be this first key that the client could not subscribe to (so the client knows all keys sent after that one were not processed).
+>
+> If the client successfully subscribes to a key, or is already subscribed to a requested key, that key MUST appear in a `RPL_METADATASUBOK` reply numeric.
+>
+> If the key's name is invalid, the server sends a `FAIL METADATA INVALID_KEY` reply to the client and continues processing keys.
+>
+> If the client does not have permission to view a given key, the server sends a `FAIL METADATA KEY_NO_PERMISSION` reply to the client and continues processing keys. However, the subscription MUST still be successful, and that key MUST appear in a `RPL_METADATASUBOK` reply numeric. In this case, the `FAIL METADATA KEY_NO_PERMISSION` reply serves as a warning indicating that the client will not receive `METADATA` messages about this key unless it gains the necessary (implementation defined) privileges later.
+
+Once the server is finished processing keys, it responds with:
+* zero or more of this numeric in any order: `RPL_METADATASUBOK`,
+* zero or more of these standard reply codes: `FAIL METADATA INVALID_KEY`, `FAIL METADATA KEY_NO_PERMISSION`
+* and MAY respond with one `FAIL METADATA TOO_MANY_SUBS` reply.
+
+### METADATA UNSUB
+
+    METADATA * UNSUB <key1> [<key2> ...]
+
+This subcommand lets clients unsubscribe from updates for the given keys.
+
+Servers process the given keys, and:
+
+> If the client successfully unsubscribes from a key, or is not subscribed to a requested key, that key MUST appear in a `RPL_METADATAUNSUBOK` reply numeric.
+>
+> If the key's name is invalid, the server sends a `FAIL METADATA INVALID_KEY` reply to the client and continues processing keys.
+
+Once the server is finished processing keys, it responds with:
+* zero or more of this numeric in any order: `RPL_METADATAUNSUBOK`
+* zero of more of this standard reply: `FAIL METADATA INVALID_KEY`
+
+### METADATA SUBS
+
+    METADATA * SUBS
+
+This subcommand returns which keys the client is currently subscribed to.
+
+The server responds with zero or more `RPL_METADATASUBS` numerics. The server MAY return the keys in any order. The server MUST NOT list the same key multiple times in a response to this subcommand.
+
```

Comment:
Without a batch or ending numeric, this is kinda tricky to deal with.

Reply:
Batches are a dependency of this spec